Cookie partitioning quirk doesn't handle redirects

sysrqb opened the pull request:
#### 802b4aa786a0bbf9712f358073e2b030b7c39a84
<pre>
Cookie partitioning quirk doesn&apos;t handle redirects
<a href="https://bugs.webkit.org/show_bug.cgi?id=285307">https://bugs.webkit.org/show_bug.cgi?id=285307</a>
<a href="https://rdar.apple.com/142278344">rdar://142278344</a>

Reviewed by Alex Christensen.

The assertion that the callback block is only set once for a task was not
correct. This patch adjusts the logic so it allows for redirects. The original
patch also reversed the first/third party functions, so I&apos;m fixing that here,
too.

* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::setCookieTransformForThirdPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransformForFirstPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransform):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):

Canonical link: <a href="https://commits.webkit.org/288399@main">https://commits.webkit.org/288399@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd951715d9d0cdcbc45ec9a00a1168c6017a0db4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/83025 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/2666 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/37318 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/88134 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/34071 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/2741 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/10586 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/64651 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/22411 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/86080 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2006 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/75498 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/44932 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/1913 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/33105 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73105 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/30392 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/89500 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/10307 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/7454 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73083 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/10536 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/71313 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/72304 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/16479 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/15223 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/1686 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/12841 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/10260 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/15772 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/10132 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/13597 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/11901 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -142,18 +142,22 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
+    ASSERT(!task()._cookieTransformCallback || isRedirect == IsRedirect::Yes);
+#else
+    UNUSED_PARAM(isRedirect);
+#endif
```
Nit - this can be changed to:
```suggestion
    ASSERT_UNUSED(isRedirect, !task()._cookieTransformCallback || isRedirect == IsRedirect::Yes);
```

rr-codes commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -142,18 +142,22 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
         return;
 
+#ifndef NDEBUG
+    ASSERT(!task()._cookieTransformCallback || isRedirect == IsRedirect::Yes);
```
I think this can just use `ASSERT_UNUSED` instead of having to use ifdefs right

sysrqb replied:
I probably knew this at some point :)

achristensen07 commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -142,18 +142,18 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
-void NetworkTaskCocoa::setCookieTransformForFirstPartyRequest(const WebCore::ResourceRequest& request)
+void NetworkTaskCocoa::setCookieTransformForThirdPartyRequest(const WebCore::ResourceRequest& request, IsRedirect isRedirect)
 {
     if (!request.isThirdParty())
```
It seems like this should be an assertion, and it seems like NetworkTaskCocoa::setCookieTransformForFirstPartyRequest should have a complimentary assertion.

achristensen07 commented on Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
```diff
@@ -142,18 +142,18 @@ static bool shouldCapCookieExpiryForThirdPartyIPAddress(const WebCore::IPAddress
 
+    ASSERT_UNUSED(isRedirect, !task()._cookieTransformCallback || isRedirect == IsRedirect::Yes);
+    if (!!task()._cookieTransformCallback)
+        task()._cookieTransformCallback = nil;
```
The !! is probably not necessary.  It also seems like this could just be an unconditional setting of the callback to nil, which would be equal or fewer ObjC calls.

achristensen07 reviewed (approved):
Can this be tested?

Commit:
Cookie partitioning quirk doesn't handle redirects
https://bugs.webkit.org/show_bug.cgi?id=285307
rdar://142278344

Reviewed by Alex Christensen.

The assertion that the callback block is only set once for a task was not
correct. This patch adjusts the logic so it allows for redirects. The original
patch also reversed the first/third party functions, so I'm fixing that here,
too.

* Source/WebKit/NetworkProcess/cocoa/NetworkDataTaskCocoa.mm:
(WebKit::NetworkDataTaskCocoa::NetworkDataTaskCocoa):
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkTaskCocoa.mm:
(WebKit::NetworkTaskCocoa::setCookieTransformForThirdPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransformForFirstPartyRequest):
(WebKit::NetworkTaskCocoa::setCookieTransform):
(WebKit::NetworkTaskCocoa::willPerformHTTPRedirection):

Canonical link: https://commits.webkit.org/288399@main
